### ISSUE TITLE
[no-unused-var] False positive when declaring interface in namespace with the same name

### ISSUE THREAD

**Issue by Guillaume-Mayer** [typescript-eslint/typescript-eslint#2595]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**
*index.d.ts*
```ts
declare namespace A {
  export interface A {
    foo: string;
  }
}
```

**Expected Result**

No warning or a warning if the namespace is never used.

**Actual Result**

Give me this warning: **'A' is defined but never used.**, on the namespace (even if it's used).

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.2.0` |
| `@typescript-eslint/parser`        | `4.2.0` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.9.0` |
| `node`                             | `12.18.4` |

**Config**

```yaml
parser: "@typescript-eslint/parser"

extends:
- eslint:recommended
- plugin:@typescript-eslint/recommended
- prettier/@typescript-eslint
- plugin:prettier/recommended

parserOptions:
  ecmaVersion: 2019
  sourceType: module
  project: ./tsconfig.json

rules:
  "@typescript-eslint/no-floating-promises": error

ignorePatterns:
- dist/
```

**Comment by bradzacher** [comment]
The bug here occurring because you have an interface with the same name as the namespace within the namespace.
ESLint is marking the interface's variable as being used instead of the module's variable.

We have passing tests for `declare`d namespaces in `.d.ts` files, but not when there are things with the same name. Those tests pass because ESLint finds no variable within the scope and then looks in the parent scope.

**Comment by mbpreble** [comment]
Have a failing test case for this one at least (added to the array of `valid` test cases in `no-unused-vars.test.ts`):

```
{
  code: `
declare namespace A {
  export interface A {
    foo: string;
  }
}
  `,
  filename: 'foo.d.ts'
}
```

As indicated by @bradzacher there's no error if the name of the interface is changed to be different from the name of the namespace.
